### PR TITLE
Add a second recovery path if BLE stack stops being able to receive adverts

### PR DESCRIPTION
**Pull request by lis0r**
Add a second recovery path if BLE stack stops being able to receive adverts

## Description:

I have twice seen the receiving of passive adverts stall, requiring a device restart to resume their reception. The issue is highly intermittent, and I have not found a means of reproducing it at will. I suspect the issue is related to https://github.com/espressif/esp-idf/issues/4001 .

The logs contain a lines stating "BLE: scan stall? no adverts > 120s, restart BLE". The restart of NimBLE is successful, meaning that Tasmota does not restart. However, the restart of NimBLE is not sufficient to rectify the underlying problem, so the message repeats every 2 minutes.

I have adjusted the check so that it will try to restart NimBLE after 120s of no advertisements. After this, if there is still no advertisements after 240s, it will restart Tasmota.

## Checklist:
  - [X] The pull request is done against the latest development branch
  - [X] Only relevant files were touched
  - [X] Only one feature/fix was added per PR and the code change compiles without warnings
  - [X] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [X] The code change is tested and works with Tasmota core ESP32 V.2.0.14
  - [X] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

### CODE REVIEW

**Comment by Jason2866** [comment]
There will be no merge. No driver is allowed to restart Tasmota just when the specific driver fails forever reason.

**Comment by s-hadinger** [comment]
There is one exception though with zigbee. If the MCU goes into an invalid state, there is no choice other than a full restart of both Tasmota and the MCU. 

I'm not familiar with BLE. What does not adverts mean? Can it happen if there are no BLE devices in range?

**Comment by barbudor** [comment]
I assume this also occurs only when using active scan?
I'm all passive with PVVX'ed Mija and I haven't seen this blocking

Probably can also be done with a rule

**Comment by lis0r** [comment]
It is definitely not a case of no devices in range - my house is a BLE heavy environment. The node in question normally reports 13 devices, and there are more devices just slightly out of range. From what I can see, the reason it finds zero devices is because the BLE controller has taken an error so severe that even restarting the BLE stack is insufficient to fix it. Manually restarting Tasmota quickly and definitively fixes the BLE stack.

I have active scans disabled to save power, and I operate purely on receiving passive adverts from a number of MJ_HT_V1,  MCCGQ02HL, and SJWS01LM devices running their stock firmware. As mentioned, it's highly intermittent - I've been running otherwise OK for months, but it's happened twice within the last few days, with no recent changes to set up or radio environment.

It should be noted that the BLE driver can already restart Tasmota: if restarting NimBLE actively fails or times out, Tasmota will be restarted after 10 seconds. This change is effectively building upon that model to cope with a false positive.

I will investigate rules.

**Comment by s-hadinger** [comment]
Let me rephrase my question. What happens if a user enables the driver but has zero BLE device in range? could Tasmota also reboot every 240 seconds?

**Comment by Jason2866** [comment]
> It should be noted that the BLE driver can already restart Tasmota: if restarting NimBLE actively fails or times out, Tasmota will be restarted after 10 seconds. This change is effectively building upon that model to cope with a false positive.

This part of the code should not be there. Seems it was overlooked in the review.

**Comment by Jason2866** [comment]
Iirc this driver is not ready for Arduino core 3.0.0 since it uses h2zero `NimBLE-Arduino` library. For Arduino core 3.0.0 to support BLE with C2 and C6 too it is needed to switch to h2zero `esp-nimble-cpp`. To use this library some changes are needed.

**Comment by lis0r** [comment]
Berry bodge around, for others looking for a similar work around:

```
last_ads = 0
def ble_trigger(value, trigger, msg)
  if value['adverts'] == last_ads
    tasmota.cmd("restart 1")
  end
  last_ads = value['adverts']
end
tasmota.add_rule("BLE", ble_trigger)
```
